src/Pipeline/Pipeline.php

```diff
@@ -119,124 +147,252 @@ protected function getConnection()
+     */
+    protected function onResponseError(ConnectionInterface $connection, CommandInterface $command, ErrorResponseInterface $response)
+    {
+        if ($this->throwOnErrorResponse) {
+            throw new ServerException($response->getMessage());
         }
 
-        return $responses;
+        return $response;
     }
 
     /**
-     * Flushes the buffer holding all of the commands queued so far.
+     * Handles traversable RESP array (prefix `*`) responses returned by Redis.
      *
-     * @param bool $send Specifies if the commands in the buffer should be sent to Redis.
+     * @param ConnectionInterface $connection Redis connection that returned the error
+     * @param CommandInterface    $command    Command affected by the error
+     * @param MultiBulkIterator   $response   Traversable response instance
+     *
+     * @return iterable
+     */
+    protected function onResponseTraversable(ConnectionInterface $connection, CommandInterface $command, MultiBulkIterator $response): iterable
+    {
+        if (!$this->traversable) {
+            $response = iterator_to_array($response, false);
+        }
+
+        return $response;
+    }
+
+    /**
+     * Flushes the current pipeline queue.
      *
      * @return $this
      */
-    public function flushPipeline($send = true)
+    public function flushQueued(): self
     {
-        if ($send && !$this->pipeline->isEmpty()) {
-            $responses = $this->executePipeline($this->getConnection(), $this->pipeline);
-            $this->responses = array_merge($this->responses, $responses);
-        } else {
-            $this->pipeline = new \SplQueue();
+        if ($this->traversable) {
+            throw new PipelineException($this, sprintf(
+                '%s does not support intermediate flushes when configured to return Traversable responses',
+                static::class
+            ));
         }
 
+        $this->flushQueuedInternal();
+
         return $this;
     }
 
     /**
-     * Marks the running status of the pipeline.
+     * Flushes the current pipeline queue (internal method).
      *
-     * @param bool $bool Sets the running status of the pipeline.
-     *
-     * @throws ClientException
+     * @return void
      */
-    private function setRunning($bool)
+    protected function flushQueuedInternal(): void
     {
-        if ($bool && $this->running) {
-            throw new ClientException('The current pipeline context is already being executed.');
+        $this->executing = true;
+
+        $responses = $this->executePipeline($this->getConnection());
+
+        if ($this->traversable) {
+            $this->responses = new Iterator\Responses($responses);
+        } else {
+            if ($this->responses === null) {
+                $this->responses = new AppendIterator();
+            }
+
+            $this->responses->append(new ArrayIterator(
+                iterator_to_array($responses, false)
+            ));
         }
```

Comment:
Apologies if my comment seem like nitpicks, I'll mostly be making suggestions to improve readability

```php
// since the else isn't exactly connected to the logic here, we ca break them apart and it'll still
// be comfortable for the reader
if ($this->traversable) {
    $this->responses = new Iterator\Responses($responses);
    return;
}

if ($this->responses === null) {
    $this->responses = new AppendIterator();
}
$this->responses->append(new ArrayIterator(
    iterator_to_array($responses, false)
));

```

Reply:
Makes sense!